framework/src/meshgenerators/CircularBoundaryCorrectionGenerator.C

```diff
@@ -0,0 +1,426 @@
+  // other.
+  // Step 1: find the farthest point from the first point
+  Real tmp_dis(0.0);
+  unsigned int farthest_pt_id(0);
+  for (unsigned int i = 1; i < pts_list.size(); i++)
+    if ((pts_list[i] - pts_list[0]).norm() > tmp_dis)
+    {
+      tmp_dis = (pts_list[i] - pts_list[0]).norm();
+      farthest_pt_id = i;
+    }
+  // Step 2: find the third point that is nearly in the middle of the first two
+  unsigned int mid_pt_id(0);
+  for (unsigned int i = 1; i < pts_list.size(); i++)
+    if (i != farthest_pt_id && std::abs((pts_list[farthest_pt_id] - pts_list[i]).norm() -
+                                        (pts_list[0] - pts_list[i]).norm()) < tmp_dis)
+    {
+      tmp_dis = std::abs((pts_list[farthest_pt_id] - pts_list[i]).norm() -
+                         (pts_list[0] - pts_list[i]).norm());
+      mid_pt_id = i;
+    }
+
+  const Real x12 = pts_list[0](0) - pts_list[farthest_pt_id](0);
+  const Real x13 = pts_list[0](0) - pts_list[mid_pt_id](0);
+
+  const Real y12 = pts_list[0](1) - pts_list[farthest_pt_id](1);
+  const Real y13 = pts_list[0](1) - pts_list[mid_pt_id](1);
+
+  const Real sx13 =
+      pts_list[0](0) * pts_list[0](0) - pts_list[mid_pt_id](0) * pts_list[mid_pt_id](0);
+  const Real sy13 =
+      pts_list[0](1) * pts_list[0](1) - pts_list[mid_pt_id](1) * pts_list[mid_pt_id](1);
+
+  const Real sx21 =
+      pts_list[farthest_pt_id](0) * pts_list[farthest_pt_id](0) - pts_list[0](0) * pts_list[0](0);
+  const Real sy21 =
+      pts_list[farthest_pt_id](1) * pts_list[farthest_pt_id](1) - pts_list[0](1) * pts_list[0](1);
+
+  const Real f =
+      (sx13 * x12 + sy13 * x12 + sx21 * x13 + sy21 * x13) / (2.0 * (-y13 * x12 + y12 * x13));
+  const Real g =
+      (sx13 * y12 + sy13 * y12 + sx21 * y13 + sy21 * y13) / (2.0 * (-x13 * y12 + x12 * y13));
+
+  const Real c = -pts_list[0](0) * pts_list[0](0) - pts_list[0](1) * pts_list[0](1) -
+                 2.0 * g * pts_list[0](0) - 2.0 * f * pts_list[0](1);
+  const Real r2 = f * f + g * g - c;
+
+  radius = std::sqrt(r2);
+  const Point circ_origin = Point(-g, -f, 0.0);
+
+  for (const auto & pt : pts_list)
+    if (!MooseUtils::absoluteFuzzyEqual((pt - circ_origin).norm(), radius, tol))
+      paramError("input_mesh_circular_boundaries", "the boundary provided is not circular.");
+  return circ_origin;
+}
+
+Real
+CircularBoundaryCorrectionGenerator::generalCirCorrFactor(
+    const std::vector<std::pair<Point, Point>> & bd_side_list,
+    const Point & circle_center,
+    const bool is_closed_loop,
+    Real & c_coeff,
+    Real & end_node_disp) const
+{
+  Real acc_area(0.0);
+  Real acc_azi(0.0);
+  std::vector<Real> d_azi_list;
+  for (const auto & bd_side : bd_side_list)
+  {
+    const Point v1 = bd_side.first - circle_center;
+    const Point v2 = bd_side.second - circle_center;
+    // Use cross to calculate r * r * sin(d_azi_i)
+    acc_area += v1.cross(v2).norm() / v1.norm() / v2.norm();
```

Comment:
this seems normalized by r^2 ? so just sin(d_azi_i)

Reply:
Yes, the comment will be revised to avert confusion